src/messages/embed.ts

```diff
@@ -0,0 +1,257 @@
+import type {
+	APIEmbed,
+	APIEmbedAuthor,
+	APIEmbedField,
+	APIEmbedFooter,
+	APIEmbedImage,
+	APIEmbedProvider,
+	APIEmbedThumbnail,
+	APIEmbedVideo,
+} from 'discord-api-types/v8';
+
+/**
+ * Represents an embed in a message (image/video preview, rich embed, etc.)
+ */
+export class Embed implements APIEmbed {
+	/**
+	 * Embed Fields.
+	 */
+	public fields: APIEmbedField[] = [];
+
+	/**
+	 * The embed title.
+	 */
+	public title?: string;
+
+	/**
+	 * The embed description.
+	 */
+	public description?: string;
+
+	/**
+	 * The embed url.
+	 */
+	public url?: string;
+
+	/**
+	 * The embed bar color.
+	 */
+	public color?: number;
+
+	/**
+	 * The timestamp of the embed.
+	 */
+	public timestamp?: string;
+
+	/**
+	 * The embed thumbnail data.
+	 */
+	public thumbnail?: APIEmbedThumbnail;
+
+	/**
+	 * The embed image data.
+	 * @since 0.0.1
+	 */
+	public image?: APIEmbedImage;
+
+	/**
+	 * Received video data.
+	 * @since 0.0.1
+	 */
+	public video?: APIEmbedVideo;
+
+	/**
+	 * The embed author data.
+	 * @since 0.0.1
+	 */
+	public author?: APIEmbedAuthor;
+
+	/**
+	 * Received data about the embed provider.
+	 * @since 0.0.1
+	 */
+	public provider?: APIEmbedProvider;
+
+	public footer?: APIEmbedFooter;
+
+	public constructor(data: APIEmbed = {}) {
+		this.title = data.title;
+		this.description = data.description;
+		this.url = data.url;
+		this.color = data.color;
+		this.thumbnail = data.thumbnail;
+		this.image = data.image;
+		this.video = data.video;
+		this.author = data.author;
+		this.provider = data.provider;
+		this.footer = data.footer;
+
+		if (data.timestamp) this.timestamp = new Date(data.timestamp).toISOString();
+		if (data.fields) this.fields = data.fields;
+	}
+
+	/**
+	 * The date displayed on this embed
+	 */
+	public get createdAt(): Date | null {
+		return this.timestamp ? new Date(this.timestamp) : null;
+	}
+
+	/**
+	 * The accumulated length for the embed title, description, fields, footer text, and author name
+	 */
+	public get length(): number {
+		return (
+			(this.title?.length ?? 0) +
+			(this.description?.length ?? 0) +
+			(this.fields.length >= 1
+				? this.fields.reduce((prev, curr) => prev + curr.name.length + curr.value.length, 0)
+				: 0) +
+			(this.footer?.text.length ?? 0) +
+			(this.author?.name?.length ?? 0)
+		);
+	}
+
+	/**
+	 * Adds a field to the embed (max 25).
+	 * @param name The name of this field
+	 * @param value The value of this field
+	 * @param inline If this field will be displayed inline
+	 */
+	public addField(name: string, value: string, inline = false): this {
+		return this.addFields({ name, value, inline });
```

Comment:
This (or the `addFields` method) should check if there are already 25 fields (not sure if an error should be thrown in the case). Also missing input validation for the data.
```suggestion
if (this.fields.length === 25) return this;
		return this.addFields({ name, value, inline });
```